rfd/0095-okta-integration.md

```diff
@@ -0,0 +1,405 @@
+endpoint: https://my-org-endpoint.okta.com
+api-token: API-TOKEN-FROM-OKTA # https://developer.okta.com/docs/guides/create-an-api-token/main/
+```
+
+#### Mapping between Teleport users and Okta users
+
+If the provider is configured with `use_uid_mapping` to `false`, which is the default, the Okta
+provider will attempt to look for users in Okta using the user's Teleport username. If this is
+set to true, provider will look for a provider config `okta_user_id ` to determine which user ID
+to use for Okta user lookups.
+
+```yaml
+kind: user
+version: v2
+metadata:
+  name: alice
+spec:
+  roles: ['devs']
+  traits:
+    logins: ['admin']
+    kubernetes_groups: ['edit']
+  provider_configs:
+    okta:prod:
+      # Can be any user ID supported by Okta, documented in
+      # https://developer.okta.com/docs/reference/api/users/#get-user
+      okta_user_id: alice@mydomain.com
+```
+
+#### Launching Okta applications
+
+[Application objects from Okta](https://developer.okta.com/docs/reference/api/apps/#application-object) can return one or more `appLinks` which, when launched directly from
+Teleport, will forward to the expected application along with Okta authentication. These links
+will be used as the application URLs synced from Okta.
+
+**Note**: These links will be redirected to their actual external URLs rather than living behind
+Teleport's proxy.
+
+### Backend modifications
+
+A few modifications to existing services, databases, and models will be necessary to support this
+functinoality.
+
+#### Applications
+
+Applications will need some new information to support this:
+
+* A new flag to indicate that the application has been externally synchronized. This will allow
+  Teleport to distinguish between regular Teleport applications and external applications.
+* User specific application metadata for external applications. This will dictate to the UI what
+  kind of options to display and whether the URLs are accessible. These will be joined with the
+  application objects during query. To start, the only necessary fields for this object are a
+  user mapping, a corresponding external identity provider name, and a boolean flag as to
+  whether the user has access.
+
+#### External groups
+
+A new notion of external groups will need to be added. This represent external groups known to
+an identity provider. Additionally, similar to applications, there must be a notion of user
+specific group metadata that contains information as to whether a user is a member of the group.
+
+#### Users
+
+Users will need new provider specific configuration fields. These will necessarily be pretty
+free-form.
+
+#### Request access service
+
+The requests for external identity providers will be longer term than Teleport's current notion
+of requests. The service will either need to be expanded to support long term requests or
+turned into a separate service that supports access requests that cannot be later dropped.
+
+### Security
```

Comment:
We should also address audit logging somewhere in this RFD. What new audit events will we emit?